### PR TITLE
[HZ-566] SQL: CREATE INDEX implementation

### PR DESCRIPTION
**Pull request by Fly-Style**
`CREATE INDEX` was fully implemented.
Grammar : 
```
CREATE INDEX [ IF NOT EXISTS ] name ON map_name ( column_name [, ...] )
[ TYPE ( SORTED | HASH | BITMAP ) ]  
[ OPTIONS ( 'option_name' = 'option_value' [, ...] ) ]
```
Options are available only for BITMAP index since it has additional `BitmapIndexConfig`. Those options are supported:
- `unique_key`
- `unique_key_transformation`.

In case of SORTED/HASH index, options usage causes `QueryException`.

`DROP INDEX` _syntax_ was supported, but not the execution path (`DROP INDEX` query throws `UnsupportedException`).

Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [x] Label `Add to Release Notes` or `Not Release Notes content` set
- [x] Request reviewers if possible

### CODE REVIEW

File: hazelcast-sql/src/main/codegen/includes/parserImpls.ftl [hunk 1]
```diff
@@ -232,6 +232,101 @@ SqlDrop SqlDropMapping(Span span, boolean replace) :
...
+
+    SqlIdentifier name;
+    boolean ifExists = false;
+}
+{
+    <INDEX>
+    [
+        <IF> <EXISTS> { ifExists = true; }
+    ]
+    name = CompoundIdentifier()
```
Comment (viliam-durina) [inline hazelcast-sql/src/main/codegen/includes/parserImpls.ftl:324]: This must be SimpleIdentifier()

File: hazelcast-sql/src/main/codegen/includes/parserImpls.ftl [hunk 2]
```diff
@@ -232,6 +232,101 @@ SqlDrop SqlDropMapping(Span span, boolean replace) :
...
+    {
```
Comment (viliam-durina) [inline hazelcast-sql/src/main/codegen/includes/parserImpls.ftl:325]: DROP index must also have the `ON` clause, see the TDD.

**Review by viliam-durina — CHANGES_REQUESTED** [review]

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/SqlCreateIndexTest.java [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+        checkPlan(false, true, selectSql);
+
+        instance().getSql().execute(sql);
+
+        checkPlan(true, true, selectSql);
+    }
+
+    @Test
+    public void indexArgsCannotBeDuplicatedTest() {
```
Comment (viliam-durina) [inline hazelcast-sql/src/test/java/com/hazelcast/jet/sql/SqlCreateIndexTest.java:120]: We don't test duplicate options, and those don't fail if I'm not mistaken.

**Review by viliam-durina — CHANGES_REQUESTED** [review]

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/SqlCreateIndexTest.java (hunk 3 above)
Comment (Fly-Style) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/SqlCreateIndexTest.java:120]: For `map.addIndex(IndexType.SORTED, "this", "this");` : 

```
java.lang.IllegalArgumentException: Duplicate attribute name [attributeName=this, indexConfig=IndexConfig{name=null, type=SORTED, attributes=[this, this]}]
```

Comment (Fly-Style) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/SqlCreateIndexTest.java:120]: It means, both test and check are makes sense.

Comment (viliam-durina) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/SqlCreateIndexTest.java:120]: I mean duplicate OPTIONS, not attributes.

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/parse/HazelcastSqlParserTest.java [hunk 4]
```diff
@@ -82,6 +83,114 @@ public void test_createMapping(boolean replace, boolean ifNotExists) throws SqlP
...
+    public void test_createIndexRequiresIndexType(boolean ifNotExists) {
+        // given
+        String sql = "CREATE INDEX "
+                + (ifNotExists ? "IF NOT EXISTS " : "")
+                + "index_name "
+                + "ON mapping_name "
+                + "(column_name1, column_name2);";
+
+        // when & then
+        assertThatThrownBy(() -> parse(sql))
```
Comment (viliam-durina) [inline hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/parse/HazelcastSqlParserTest.java:185]: The test fails due to the unallowed `;` at the end. TYPE is optional.

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/SqlCreateIndexTest.java [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+        String sql = "CREATE INDEX IF NOT EXISTS idx ON " + MAP_NAME + " (this, this) TYPE BITMAP";
+        assertThatThrownBy(() -> instance().getSql().execute(sql))
+                .hasMessageContaining("specified more than once");
+    }
+
+    @Test
+    public void indexExistsTest() {
```
Comment (viliam-durina) [inline hazelcast-sql/src/test/java/com/hazelcast/jet/sql/SqlCreateIndexTest.java:126]: There should also be inverse test: that the IF NOT EXISTS does nothing, if the index already exists.

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/SqlCreateIndexTest.java (hunk 3 above)
Comment (Fly-Style) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/SqlCreateIndexTest.java:120]: Ah, sorry. You pointed to index**Args**CannotBeDuplicatedTest(), and I immediately switched to args context :)

**Review by viliam-durina — CHANGES_REQUESTED** [review]

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/SqlPlanImpl.java [hunk 6]
```diff
@@ -193,6 +194,140 @@ public SqlResult execute(QueryId queryId, List<Object> arguments, long timeout)
...
+        }
+
+        @Override
+        public boolean isPlanValid(PlanCheckContext context) {
+            return true;
+        }
+
+        @Override
+        public void checkPermissions(SqlSecurityContext context) {
+            context.checkPermission(new SqlPermission(name, ACTION_CREATE));
```
[the remark this answers is not in the document]
Comment (Fly-Style) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/SqlPlanImpl.java:263]: I planned to do it as follow-up PR, WDYT?

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/SqlCreateIndexTest.java (hunk 3 above)
Comment (viliam-durina) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/SqlCreateIndexTest.java:119]: I always add comments to some code line. Comments added to the whole PR cannot be marked as resolved.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/SqlPlanImpl.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (viliam-durina) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/SqlPlanImpl.java:263]: But the ACTION_INDEX should be in this PR, shouldn't it?

**Review by viliam-durina — APPROVED** [review]

**Pull request merged**